### PR TITLE
SparseFileTracker.Gap should keep a reference to the corresponding Range

### PR DESCRIPTION
**Pull request by tlrx**
Relates https://github.com/elastic/elasticsearch/pull/58477#discussion_r444791383

### CODE REVIEW

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/SparseFileTracker.java [hunk 1]
```diff
@@ -255,18 +255,21 @@ private long computeLengthOfRanges() {
         }
     }
 
-    private void onGapSuccess(final long start, final long end) {
+    private boolean assertPendingRangeExists(Range range) {
+        assert Thread.holdsLock(mutex);
+        final SortedSet<Range> existingRanges = ranges.tailSet(range);
+        assert existingRanges.isEmpty() == false;
+        final Range existingRange = existingRanges.first();
+        assert existingRange.start == range.start && existingRange.end == range.end && existingRange.isPending();
```
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/SparseFileTracker.java:263]: Can we `assert existingRange == range`? I think it should always be the very same object here.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/SparseFileTracker.java [hunk 2]
```diff
@@ -255,18 +255,21 @@ private long computeLengthOfRanges() {
...
+        return true;
+    }
+
+    private void onGapSuccess(final Range existingRange) {
```
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/SparseFileTracker.java:267]: Naming suggestion (here and elsewhere):

```suggestion
    private void onGapSuccess(final Range gapRange) {
```

**Review by DaveCTurner — COMMENTED** [review]
Looks good, I left two small suggestions.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/SparseFileTracker.java (hunk 1 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/SparseFileTracker.java:263]: I think we can, yes

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/SparseFileTracker.java (hunk 2 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/SparseFileTracker.java:267]: That's better, thanks

**Comment by tlrx** [comment]
Thanks David. I updated the code.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by tlrx** [comment]
Thanks David

**Pull request merged**